lib/puppet/util/character_encoding.rb

```diff
@@ -13,12 +12,16 @@ class << self
     # @raise [Puppet::Error] a failure to legitimately set external encoding or
     #   transcode string
     def convert_to_utf_8(string)
-      return string if string.encoding == Encoding::UTF_8
-
-      value_to_encode = string.dup
+      return string if (string.encoding == Encoding::UTF_8 && string.valid_encoding?)
 
       begin
-        if valid_utf_8?(value_to_encode)
+        if already_invalid_utf_8?(string)
+          # If a string is currently believed to be UTF-8, but is also not
+          # valid_encoding?, we have no recourse but to fail because we have no
+          # idea what encoding this string originally came from where it *was*
+          # valid - all we know is it's not currently valid UTF-8.
+          raise EncodingError
```

Comment:
I tried to get some numbers on `valid_encoding?` and at first it seemed like basically a no-op compared with the other operations... But I suppose it adds up on multiple checks. For example this string:
`"foo bar baz f dsa fs a fsa fsd afdsa fdas fdsa fa a afdsa afa  \xfd\xf1 \u2603"` which contains both invalid byte sequences and valid unicode, running `valid_encoding?` 10,000,000 times takes about 1 second. Calling it 3x more times is 3 more seconds, so might as well as cut it back if we can!

> If we wanted to simplify the logic, we could do something like:
> begin
  string.encode!(Encoding::UTF_8)
rescue EncodingError => e
  original_encoding = string.encoding
  string.force_encoding(Encoding::UTF_8)
  if string.valid_encoding?
    string
  else
    string.force_encoding(original_encoding)
    raise e
  end
end

I actually want to avoid doing the `encode!` first because due to collisions we can end up transcoding a string (into another valid utf-8 string) when it was actually unicode bits to begin with. Part of the heuristic is to avoid transcoding what would already be valid utf-8 bits on the assumption that they started correct. For example, this character:

ᚠ (rune - 225, 154, 160) is valid UTF-8, but if read in on a system currently running in LATIN-1, will transcode cleanly to different UTF-8, but that's likely not what we want.
```
[21] pry(main)> s = [225, 154, 160].pack('C*').force_encoding(Encoding::ISO_8859_1)
=> "\xE1\x9A\xA0"
[22] pry(main)> s.encode!(Encoding::UTF_8)
=> "á\u009A "
[23] pry(main)> s.valid_encoding?
=> true
[24] pry(main)> s.bytes
=> [195, 161, 194, 154, 194, 160]
```

 (I actually added a test for it, here: https://github.com/puppetlabs/puppet/blob/master/spec/unit/util/character_encoding_spec.rb#L13-L28)